File: _debug/redesign/v6-filter/verify.py
import json, subprocess, time, urllib.request, sys, base64
from pathlib import Path

ROOT = Path(r"E:\Program\Firefly")
NODE = r"C:\Users\lenovo\AppData\Local\Programs\kimi-desktop\resources\resources\runtime\node"
EDGE = r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
OUT = ROOT / "_debug" / "redesign" / "v6-filter"
OUT.mkdir(parents=True, exist_ok=True)

import websocket

server = subprocess.Popen([NODE, "node_modules/astro/bin/astro.mjs", "preview", "--port", "4322", "--host", "127.0.0.1"],
    cwd=ROOT, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
edge = None
results = []
def check(name, cond, detail=""):
    results.append(bool(cond))
    print(("PASS" if cond else "FAIL"), name, detail)

try:
    for i in range(60):
        try:
            urllib.request.urlopen("http://127.0.0.1:4322/", timeout=2); break
        except Exception: time.sleep(1)
    else:
        raise RuntimeError("preview server not ready")
    print("server ready")

    edge = subprocess.Popen([EDGE, "--headless=new", "--disable-gpu", "--remote-debugging-port=9224",
        "--remote-allow-origins=*", "--window-size=1440,900",
        "--user-data-dir=" + str(ROOT / "_debug" / ".edge-profile-v6b"), "about:blank"],
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    ws_url = None
    for i in range(20):
        try:
            tabs = json.loads(urllib.request.urlopen("http://127.0.0.1:9224/json", timeout=2).read())
            page = [t for t in tabs if t.get("type") == "page"]
            if page:
                ws_url = page[0]["webSocketDebuggerUrl"]; break
        except Exception: pass
        time.sleep(1)
    if not ws_url: raise RuntimeError("no CDP target")

    ws = websocket.create_connection(ws_url, timeout=20)
    mid = [0]
    def send(method, params=None):
        mid[0] += 1
        ws.send(json.dumps({"id": mid[0], "method": method, "params": params or {}}))
        deadline = time.time() + 20
        while time.time() < deadline:
            msg = json.loads(ws.recv())
            if msg.get("id") == mid[0]:
                return msg.get("result", {})
        raise TimeoutError(method)
    def js(expr):
        r = send("Runtime.evaluate", {"expression": expr, "returnByValue": True, "awaitPromise": True})
        res = r.get("result", {})
        return res.get("value")
    def shot(name):
        data = send("Page.captureScreenshot", {"format": "png"})["data"]
        (OUT / name).write_bytes(base64.b64decode(data))

    send("Page.enable"); send("Runtime.enable")
    send("Page.navigate", {"url": "http://127.0.0.1:4322/"})
    time.sleep(4)

    url0 = js("location.pathname")
    n_all = js("document.querySelectorAll('#post-list-container > *').length")
    print("initial cards:", n_all, "url:", url0)

    # 1. click filter
    js("document.getElementById('ai-filter-btn').click()")
    time.sleep(3)
    check("点击后 URL 不变", js("location.pathname") == url0, f"{url0} -> {js('location.pathname')}")
    n_filtered = js("document.querySelectorAll('#post-list-container > *').length")
    check("过滤后数量减少", isinstance(n_filtered, int) and 0 < n_filtered < n_all, f"all={n_all} filtered={n_filtered}")
    check("按钮激活态", js("document.getElementById('ai-filter-btn').hasAttribute('data-active')") is True)
    shot("01-filtered-light.png")

    # 2. click again -> restore
    js("document.getElementById('ai-filter-btn').click()")
    time.sleep(3)
    check("再点恢复全部", js("document.querySelectorAll('#post-list-container > *').length") == n_all)
    check("恢复后按钮非激活", js("document.getElementById('ai-filter-btn').hasAttribute('data-active')") is False)
    shot("02-restored-light.png")

    # 3. activate then full reload -> persist
    js("document.getElementById('ai-filter-btn').click()")
    time.sleep(3)
    send("Page.navigate", {"url": "http://127.0.0.1:4322/"})
    time.sleep(5)
    n_reload = js("document.querySelectorAll('#post-list-container > *').length")
    active3 = js("document.getElementById('ai-filter-btn').hasAttribute('data-active')")
    check("刷新后保持过滤", n_reload == n_filtered and active3 is True, f"reload={n_reload} expect={n_filtered}")

    # 4. swup: archive -> home, filter persists
    js("document.querySelector('.category-pill[data-category-name=\"__archive__\"]').click()")
    time.sleep(4)
    check("Swup 到归档页", "archive" in (js("location.pathname") or ""), js("location.pathname"))
    js("document.querySelector('.category-pill[data-category-name=\"\"]').click()")
    time.sleep(4)
    n_swup = js("document.querySelectorAll('#post-list-container > *').length")
    active4 = js("document.getElementById('ai-filter-btn') && document.getElementById('ai-filter-btn').hasAttribute('data-active')")
    check("Swup 回首页过滤保持", n_swup == n_filtered and active4 is True, f"n={n_swup} url={js('location.pathname')}")

    # 5. dark mode
    js("localStorage.setItem('theme','dark'); document.documentElement.classList.add('dark');")
    time.sleep(1)
    shot("03-filtered-dark.png")
    js("document.getElementById('ai-filter-btn').click()")
    time.sleep(3)
    check("暗色下恢复全部", js("document.querySelectorAll('#post-list-container > *').length") == n_all)
    shot("04-restored-dark.png")

    print("SUMMARY:", f"{sum(results)}/{len(results)} passed")
    ws.close()
finally:
    if edge: subprocess.run(["taskkill", "//F", "//T", "//PID", str(edge.pid)], capture_output=True)
    subprocess.run(["taskkill", "//F", "//T", "//PID", str(server.pid)], capture_output=True)
    time.sleep(2)
    import socket
    s = socket.socket()
    try:
        s.connect(("127.0.0.1", 4322)); print("WARN: port 4322 still open")
    except Exception:
        print("port 4322 closed")
    s.close()

File: _debug/redesign/v5-nav/verify.mjs
import { spawn } from "node:child_process";
import { writeFileSync } from "node:fs";

const EDGE = "C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe";
const BASE = "http://localhost:4322";
const OUT = "E:\\Program\\Firefly\\_debug\\redesign\\v5-nav";
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const edge = spawn(EDGE, [
	"--headless=new", "--disable-gpu", "--remote-debugging-port=9333",
	"--window-size=1440,900", "about:blank",
], { stdio: "ignore" });

async function getWsUrl() {
	for (let i = 0; i < 30; i++) {
		try {
			const res = await fetch("http://localhost:9333/json");
			const list = await res.json();
			const page = list.find((t) => t.type === "page");
			if (page) return page.webSocketDebuggerUrl;
		} catch {}
		await sleep(500);
	}
	throw new Error("no CDP");
}

const ws = new WebSocket(await getWsUrl());
await new Promise((r) => (ws.onopen = r));
let id = 0;
const pending = new Map();
ws.onmessage = (e) => {
	const msg = JSON.parse(e.data);
	if (msg.id && pending.has(msg.id)) { pending.get(msg.id)(msg); pending.delete(msg.id); }
};
function send(method, params = {}) {
	return new Promise((resolve) => {
		const mid = ++id;
		pending.set(mid, resolve);
		ws.send(JSON.stringify({ id: mid, method, params }));
	});
}
async function evaljs(expr) {
	const r = await send("Runtime.evaluate", { expression: expr, awaitPromise: true, returnByValue: true });
	return r.result?.result?.value;
}
async function shot(name) {
	const r = await send("Page.captureScreenshot", { format: "png" });
	writeFileSync(`${OUT}\\${name}.png`, Buffer.from(r.result.data, "base64"));
	console.log("shot:", name);
}
async function goto(url) {
	await send("Page.navigate", { url });
	await sleep(2500);
}

await send("Page.enable");
await send("Emulation.setDeviceMetricsOverride", { width: 1440, height: 900, deviceScaleFactor: 1, mobile: false });

// 1. 首页：检查游戏 tab 是否隐藏、点击主页 tab 定位到主体
await goto(`${BASE}/`);
const navLinks = await evaljs(`Array.from(document.querySelectorAll('#navbar a')).map(a => a.textContent.trim() + ' => ' + a.getAttribute('href'))`);
console.log("nav links:", JSON.stringify(navLinks, null, 1));
const hasGame = await evaljs(`document.getElementById('navbar').textContent.includes('游戏')`);
console.log("游戏 tab 存在?", hasGame);

// 点击主页 tab（桌面 DropdownMenu 链接）
await evaljs(`(function(){
  const a = Array.from(document.querySelectorAll('#navbar .dropdown-container > a[href]')).find(x => x.getAttribute('href').includes('#swup-container'));
  if (a) a.click();
  return !!a;
})()`);
await sleep(2500);
const afterHomeClick = await evaljs(`JSON.stringify({scrollY: Math.round(window.scrollY), swupTop: document.getElementById('swup-container')?.getBoundingClientRect().top})`);
console.log("首页点击主页 tab 后:", afterHomeClick);
await shot("01-home-click-home-tab");

// 点击 logo 回顶部
await evaljs(`document.getElementById('nav-logo-link').click()`);
await sleep(2500);
const afterLogo = await evaljs(`Math.round(window.scrollY)`);
console.log("首页点击 logo 后 scrollY:", afterLogo);
await shot("02-home-click-logo");

// 2. 从文章页点击主页 tab
const postUrl = await evaljs(`(function(){const a=document.querySelector('#swup-container a[href*="/posts/"]');return a?a.href:null;})()`);
console.log("post url:", postUrl);
if (postUrl) {
	await goto(postUrl);
	await evaljs(`(function(){
	  const a = Array.from(document.querySelectorAll('#navbar .dropdown-container > a[href]')).find(x => x.getAttribute('href').includes('#swup-container'));
	  if (a) a.click();
	  return !!a;
	})()`);
	await sleep(3000);
	const crossPage = await evaljs(`JSON.stringify({path: location.pathname, scrollY: Math.round(window.scrollY), hash: location.hash})`);
	console.log("文章页点击主页 tab 后:", crossPage);
	await shot("03-post-click-home-tab");

	// 从文章页点击 logo
	await goto(postUrl);
	await evaljs(`document.getElementById('nav-logo-link').click()`);
	await sleep(3000);
	const crossLogo = await evaljs(`JSON.stringify({path: location.pathname, scrollY: Math.round(window.scrollY)})`);
	console.log("文章页点击 logo 后:", crossLogo);
	await shot("04-post-click-logo");
}

// 3. 移动端菜单检查（游戏 tab 隐藏 + 主页项 href）
await send("Emulation.setDeviceMetricsOverride", { width: 390, height: 844, deviceScaleFactor: 2, mobile: true });
await goto(`${BASE}/`);
const mobileLinks = await evaljs(`Array.from(document.querySelectorAll('#nav-menu-panel a')).map(a => a.textContent.trim() + ' => ' + a.getAttribute('href'))`);
console.log("mobile menu links:", JSON.stringify(mobileLinks, null, 1));
await shot("05-mobile-menu-check");

ws.close();
edge.kill();
console.log("DONE");
process.exit(0);
